modules/apps/remote-web-app/remote-web-app-web/src/main/java/com/liferay/remote/web/app/web/internal/portlet/RemoteWebAppPortlet.java

```diff
@@ -0,0 +1,154 @@
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.portlet.bridges.mvc.MVCPortlet;
+import com.liferay.portal.kernel.util.HashMapBuilder;
+import com.liferay.portal.kernel.util.ResourceBundleLoader;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+
+import java.util.Collections;
+import java.util.Dictionary;
+import java.util.Enumeration;
+import java.util.Hashtable;
+import java.util.Locale;
+import java.util.Map;
+import java.util.ResourceBundle;
+
+import javax.portlet.Portlet;
+import javax.portlet.RenderRequest;
+import javax.portlet.RenderResponse;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+
+/**
+ * @author Iván Zaera Avellón
+ */
+public class RemoteWebAppPortlet extends MVCPortlet {
+
+	public RemoteWebAppPortlet(long id, String name, String url) {
+		_id = id;
+		_name = name;
+		_url = url;
+	}
+
+	public String getName() {
+		return _name;
+	}
+
+	public synchronized void register(BundleContext bundleContext) {
+		if (_serviceRegistration != null) {
+			throw new IllegalStateException("Portlet is already registered");
+		}
+
+		Dictionary<String, Object> properties = new Hashtable<>();
+
+		properties.put(
+			"com.liferay.portlet.display-category", "category.sample");
+		properties.put("com.liferay.portlet.instanceable", true);
+		properties.put("javax.portlet.name", _getPortletName());
+		properties.put("javax.portlet.security-role-ref", "power-user,user");
+		properties.put(
+			"javax.portlet.resource-bundle", _getResourceBundleName());
+
+		_serviceRegistration = bundleContext.registerService(
+			Portlet.class, this, properties);
+
+		properties = new Hashtable<>();
+
+		properties.put("resource.bundle.base.name", _getResourceBundleName());
+		properties.put("servlet.context.name", "remote-web-app-web");
+
+		_resourceBundleLoaderServiceRegistration =
+			bundleContext.registerService(
+				ResourceBundleLoader.class,
+				locale -> _getResourceBundle(locale), properties);
+	}
+
+	@Override
+	public void render(
+		RenderRequest renderRequest, RenderResponse renderResponse) {
+
+		try {
+			PrintWriter printWriter = renderResponse.getWriter();
+
+			printWriter.print("<iframe src=\"" + _url + "\"></iframe>");
+
+			printWriter.flush();
+		}
+		catch (IOException ioException) {
+			_log.error("Unable to render HTML output", ioException);
+		}
+	}
+
+	public synchronized void unregister() {
+		if (_serviceRegistration == null) {
+			throw new IllegalStateException("Portlet is not registered");
+		}
+
+		_resourceBundleLoaderServiceRegistration.unregister();
+		_serviceRegistration.unregister();
+
+		_resourceBundleLoaderServiceRegistration = null;
+		_serviceRegistration = null;
+	}
+
+	private String _getPortletName() {
+		return "remote_web_app_" + _id;
+	}
+
+	private ResourceBundle _getResourceBundle(Locale locale) {
+
+		// TODO: this ResourceBundle needs to be truly localized (and cached)
```

Comment:
We probably need to make `name` a localized entry in the `RemoteWebAppEntry` model and then use it here.